src/plonk/circuit.rs

```diff
@@ -4,46 +4,126 @@ use std::collections::BTreeMap;
 
 use super::Error;
 use crate::arithmetic::Field;
-
 use crate::poly::Rotation;
-/// This represents a wire which has a fixed (permanent) value
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
-pub struct FixedWire(pub usize);
 
-/// This represents a wire which has a witness-specific value
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
-pub struct AdviceWire(pub usize);
+/// A column type
+pub trait ColumnType: 'static + Sized {}
+
+/// A column with an index and type
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Column<C: ColumnType> {
+    pub(crate) index: usize,
+    pub(crate) column_type: C,
+}
+
+/// An advice column
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Advice;
+
+/// A fixed column
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Fixed;
+
+/// An auxiliary column
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Aux;
+
+/// An enum over the Advice, Fixed, Aux structs
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Any {
+    /// An Advice variant
+    Advice,
+    /// A Fixed variant
+    Fixed,
+    /// An Auxiliary variant
+    Aux,
+}
+
+impl ColumnType for Advice {}
+impl ColumnType for Fixed {}
+impl ColumnType for Aux {}
+impl ColumnType for Any {}
 
-/// This represents a wire which has an externally assigned value
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
-pub struct AuxWire(pub usize);
+impl From<Column<Advice>> for Column<Any> {
+    fn from(advice: Column<Advice>) -> Column<Any> {
+        Column {
+            index: advice.index,
+            column_type: Any::Advice,
+        }
+    }
+}
+
+impl From<Column<Fixed>> for Column<Any> {
+    fn from(advice: Column<Fixed>) -> Column<Any> {
+        Column {
+            index: advice.index,
+            column_type: Any::Fixed,
+        }
+    }
+}
+
+impl From<Column<Aux>> for Column<Any> {
+    fn from(advice: Column<Aux>) -> Column<Any> {
+        Column {
+            index: advice.index,
+            column_type: Any::Aux,
+        }
+    }
+}
+
+impl From<Column<Any>> for Column<Advice> {
+    fn from(any: Column<Any>) -> Column<Advice> {
+        Column {
+            index: any.index,
+            column_type: Advice,
+        }
+    }
+}
+
+impl From<Column<Any>> for Column<Fixed> {
+    fn from(any: Column<Any>) -> Column<Fixed> {
+        Column {
+            index: any.index,
+            column_type: Fixed,
+        }
+    }
+}
+
+impl From<Column<Any>> for Column<Aux> {
+    fn from(any: Column<Any>) -> Column<Aux> {
+        Column {
+            index: any.index,
+            column_type: Aux,
+        }
+    }
+}
```

Comment:
These are fallible conversions; I suggest to use `std::convert::TryFrom` and within the conversion make sure that the `Column<Any>` struct's `column_type: Any` enum has the appropriate variant and fail otherwise.